### PR TITLE
feat: add initial storage interface

### LINKED ISSUE #13: Implement the metadata storage API on Storage Nodes
Depends on #15

### PR DESCRIPTION
**Pull request by jpcsmith**
This PR adds an initial interface for the storage node's storage. 

As most of the types are still pending, minimal, temporary types are currently used and are subject to change. 

Added issues #41 for the storage tuning.

Close #13, #14, and #15

### CODE REVIEW

File: crates/walrus-service/Cargo.toml [hunk 1]
```diff
@@ -7,12 +7,20 @@ edition = { workspace = true }
 license = { workspace = true }
 
 [dependencies]
+anyhow = "1.0.80"
```
Comment (asonnino) [inline crates/walrus-service/Cargo.toml:10]: how about eyre (https://docs.rs/eyre/latest/eyre/) it is what we use in Sui last time I checked

File: crates/walrus-service/src/storage.rs [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+        // See https://rocksdb.org/blog/2021/05/26/integrated-blob-db.html
+        options.set_enable_blob_files(true);
+
+        database.create_cf(Self::column_family_name(id), &options)?;
+
+        Ok(ShardStorage::new(id, database))
+    }
+
+    /// Stores the provided primary or secondary encoding for the given blob ID.
+    pub fn put_encoding(
```
Comment (giac-mysten) [inline crates/walrus-service/src/storage.rs]: Remark: I think references to` encoding` here should be instead `sliver`?

File: crates/walrus-service/src/storage.rs [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+        &self,
+        blob_id: &BlobId,
+        encoding: &[u8],
+        is_primary: bool,
```
Comment (giac-mysten) [inline crates/walrus-service/src/storage.rs:136]: Remark: @mlegner flagging this as we were discussing how to pass whether the encoding is primary or secondary. Should we adopt this convention everywhere? `is_primary` for the decoding/storing and `to_primary` for the encoding?

File: crates/walrus-service/src/storage.rs (hunk 2 above)
Comment (jpcsmith) [reply crates/walrus-service/src/storage.rs]: Yes, it will be updated.

File: crates/walrus-service/src/storage.rs (hunk 3 above)
Comment (jpcsmith) [reply crates/walrus-service/src/storage.rs:136]: Ehh, this is temp, id prefer the final api to be typed (e.g., an enum).

File: crates/walrus-service/Cargo.toml (hunk 1 above)
Comment (jpcsmith) [reply crates/walrus-service/Cargo.toml:10]: I had thought so as well but found that only around 16 of the crates use eyre vs 60 using anyhow.

File: crates/walrus-service/src/storage.rs (hunk 3 above)
Comment (giac-mysten) [reply crates/walrus-service/src/storage.rs:136]: Ok good! That is also what we have in the API document :)

File: crates/walrus-service/src/test_utils.rs [hunk 4]
```diff
@@ -0,0 +1,172 @@
+//! Test utilities shared between various crates.
```
Comment (mlegner) [inline crates/walrus-service/src/test_utils.rs]: Remark: As far as I can tell, this was copied more or less unchanged from scion-rs, correct?

Nit: The module docstring is not really correct as this is not shared with other crates. We could put it into a separate crate, however.

File: crates/walrus-service/src/storage.rs [hunk 5]
```diff
@@ -0,0 +1,333 @@
+#![allow(unused)]
+use std::{collections::HashSet, path::Path};
+
+use anyhow::Context;
+use rocksdb::{ColumnFamily, ColumnFamilyDescriptor, Options, DB};
+use walrus_core::BlobId;
+
+use crate::config::ShardIndex;
+
+pub type Metadata = Vec<[u8; 32]>;
```
Comment (mlegner) [inline crates/walrus-service/src/storage.rs:10]: Nit: The semantics of the metadata would be clearer with something like the following:
```suggestion
pub type Hash = [u8; 32];
pub struct Metadata {
    sliver_pair_hashes: Vec<Hash>,
}
```

File: crates/walrus-service/src/storage.rs [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+
+        Ok(Self {
+            database,
+            shards: HashSet::default(),
+        })
+    }
+
+    /// Creates storage for the specified shard, and returns a handle to its storage.
+    ///
+    /// If the storage already exists, then this simply returns the handle to the storage.
```
Comment (mlegner) [inline crates/walrus-service/src/storage.rs]: Suggestion: Either change the docstring or actually return the shard handle.

File: crates/walrus-service/src/storage.rs [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+    fn arbitrary_sliver() -> Vec<u8> {
+        vec![7; 1024]
+    }
+    fn other_sliver() -> Vec<u8> {
+        vec![201; 512]
+    }
+
+    const BLOB_ID: BlobId = [7; 32];
+    const SHARD_INDEX: ShardIndex = 17;
+    const OTHER_SHARD_INDEX: ShardIndex = 831;
```
Comment (mlegner) [inline crates/walrus-service/src/storage.rs]: Nit: Can you move the constants to the top of the tests? Would be more intuitive, in particular as `SHARDI_INDEX` is used in `empty_storage`.

File: crates/walrus-service/src/storage.rs [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+    }
+
+    impl AsRef<Storage> for TempStorage {
+        fn as_ref(&self) -> &Storage {
+            &self.storage
+        }
+    }
+
+    fn empty_storage() -> TempStorage {
+        empty_storage_with_shards(&[SHARD_INDEX])
```
Comment (mlegner) [inline crates/walrus-service/src/storage.rs:204]: Question: Why is it considered empty when there is already one shard stored?

File: crates/walrus-service/src/storage.rs [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+
+    fn arbitrary_metadata() -> Metadata {
+        (0..100u8).map(|i| [i; 32]).collect()
+    }
+
+    fn arbitrary_sliver() -> Vec<u8> {
+        vec![7; 1024]
+    }
+    fn other_sliver() -> Vec<u8> {
+        vec![201; 512]
```
Comment (mlegner) [inline crates/walrus-service/src/storage.rs:231]: Nit: When we test different lengths, I would choose one length that isn't a power of 2. Also, the sliver lengths don't necessarily need to be different by a factor of 2, we just need to make sure that the corresponding encoding rates are _at most_ 1/3 and 2/3 respectively.

File: crates/walrus-service/src/storage.rs (hunk 5 above)
Comment (jpcsmith) [reply crates/walrus-service/src/storage.rs:10]: Will change once I get the API from Karl.

File: crates/walrus-service/src/storage.rs (hunk 8 above)
Comment (jpcsmith) [reply crates/walrus-service/src/storage.rs:204]: The shard-"tables" exist, but there's no data stored.

File: crates/walrus-service/src/storage.rs (hunk 9 above)
Comment (jpcsmith) [reply crates/walrus-service/src/storage.rs:231]: I will update this once I have the code representations for the slivers, this was not meant to be representative...

File: crates/walrus-service/src/test_utils.rs (hunk 4 above)
Comment (jpcsmith) [reply crates/walrus-service/src/test_utils.rs]: Moved it to its own crate.

**Pull request merged**